### PR TITLE
Add T5 LM v1.1 encoder

### PR DESCRIPTION
**Pull request by sogartar**
The encoder shares much of the underlying stack as the decoder. Here only the encoder is presented as a class.
I have not gone out of my way to strip all decoder related stuff from the stack. Things like check-pointing and dropout are stripped.

The author attribution is added to the license of the T5 model file as this seems like a derivative work. They are both Apache 2.0.

There are a few tests of the various components and 2 tests for the entire encoder for the small and xxl variants. They relay on huggingface and the models are downloaded no the fly into the cache. The tests expect the corresponding GGUF files to be already preset and available on the file system.

### CODE REVIEW

File: sharktank/sharktank/layers/ffn_block.py [hunk 1]
```diff
@@ -22,18 +23,29 @@ class FFN(ThetaLayer):
     def __init__(
         self,
         theta: Theta,
+        is_gated: bool = None,
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/ffn_block.py:26]: How about setting a default boolean value here instead of None.

File: sharktank/sharktank/layers/ffn_block.py [hunk 2]
```diff
@@ -22,18 +23,29 @@ class FFN(ThetaLayer):
...
+        activation_fn: Optional[Callable[[AnyTensor], AnyTensor]] = None,
     ):
         super().__init__(theta)
 
-        self.add_module("ffn_gate", LinearLayer(theta("ffn_gate")))
+        self.is_gated = is_gated is None or is_gated
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/ffn_block.py:31]: If default value is boolean, will not need an `or` statement here. Just `self.is_gated = is_gated`

File: sharktank/sharktank/models/t5/t5.py [hunk 3]
```diff
@@ -0,0 +1,1136 @@
...
+)
+from ... import ops
+from ...types.theta import Theta
+from ...layers import FFN
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class T5Config:
```
Comment (archana-ramalingam) [inline sharktank/sharktank/models/t5/t5.py:35]: This might need to move to `layers/configs/t5_configs.py`. Even better if moved to `llm_configs.py`

File: sharktank/sharktank/layers/ffn_block.py (hunk 1 above)
Comment (sogartar) [reply sharktank/sharktank/layers/ffn_block.py:26]: Having the default as None helps with propagation of the underlying default value to callers. They can easily say that their default should change with the default of the underlying function.
E.g.
```
def f(x: bool = None)
    ...

def g(x: bool = None):
    f(x)
```

Comment (sogartar) [reply sharktank/sharktank/layers/ffn_block.py]: I changed the default to `True` to avoid confusion.

File: sharktank/sharktank/layers/ffn_block.py (hunk 2 above)
Comment (sogartar) [reply sharktank/sharktank/layers/ffn_block.py]: Done

File: sharktank/sharktank/models/t5/t5.py (hunk 3 above)
Comment (sogartar) [reply sharktank/sharktank/models/t5/t5.py]: I moved it there.

File: sharktank/sharktank/models/t5/t5.py [hunk 4]
```diff
@@ -0,0 +1,1067 @@
...
+        if hidden_states.dtype == torch.float16:
+            clamp_value = ops.elementwise(
+                torch.where,
+                ops.elementwise(torch.isinf, hidden_states).any(),
+                torch.finfo(hidden_states.dtype).max - 1000,
+                torch.finfo(hidden_states.dtype).max,
+            )
+            hidden_states = ops.elementwise(
+                torch.clamp, hidden_states, min=-clamp_value, max=clamp_value
+            )
```
Comment (sogartar) [inline sharktank/sharktank/models/t5/t5.py:574]: @archana-ramalingam, you mentioned some training remnants, do you mean these?
I am a bit reluctant to remove these as if the train was done with clamping, shouldn't inference also clamp? In general we want to run the same function as it was trained on.
The question is if we get a trained fp16 T5, would it have used this clamping during training?

File: .github/workflows/ci_eval.yaml [hunk 5]
```diff
@@ -123,3 +125,63 @@ jobs:
...
+          # Try with the latest IREE nightly releases, not what iree-turbine pins.
+          # We could also pin to a known working or stable version.
+          # This should eventually stabilize. Do the best we can for now.
+          pip install -f https://iree.dev/pip-release-links.html --upgrade --pre \
+            iree-base-compiler \
+            iree-base-runtime
+
+      - name: Run long running tests
+        run: |
+          pytest --longrun \
```
Comment (IanNod) [inline .github/workflows/ci_eval.yaml:184]: How long does this take to run to have this fall under longrun? Ideally we should be testing the models on presubmit if possible

File: sharktank/conftest.py [hunk 6]
```diff
@@ -256,6 +272,16 @@ def get_model_artifacts(request: FixtureRequest):
     model_path["llama3_405b_fp8_model_path"] = set_fixture_from_cli_option(
         request, "--llama3-405b-fp8-model-path", "llama3_405b_fp8_model"
     )
+    model_path["google__t5_v1_1_small_fp32_model_path"] = set_fixture_from_cli_option(
```
Comment (IanNod) [inline sharktank/conftest.py:275]: Do we need 2 separate model_path flags for small vs xxl models instead of just calling this twice from our tests?

File: sharktank/sharktank/ops/default_impls.py [hunk 7]
```diff
@@ -141,36 +141,19 @@ def elementwise_binary(operator, x, y, *args, **kwargs):
...
-    (Tensor)
-    (Tensor, Tensor)
-    (int, Tensor, Tensor)
-    ```
-    """
-    res = elementwise(operator, x, y)
-    for arg in args:
-        res = elementwise(operator, res, arg)
-    return res
+def elementwise_trenary(operator, x, y, z, *args, **kwargs):
```
Comment (IanNod) [inline sharktank/sharktank/ops/default_impls.py:150]: Why is this change needed? Also should this be elementwise_ternary?

File: sharktank/sharktank/ops/signatures.py [hunk 8]
```diff
@@ -374,6 +376,23 @@ def _gather_trampoline(
         d.fail(dispatch_args)
 
 
+def gelu_tanh_approximation(input: AnyTensor) -> AnyTensor:
```
Comment (IanNod) [inline sharktank/sharktank/ops/signatures.py:379]: lots of hardcoded values here. Are those all specified by the linked paper?

File: sharktank/tests/models/t5/t5_test.py [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+        [
+            "google/t5-v1_1-small",
+            "google/t5-v1_1-xxl",
+        ]
+    )
+    @longrun
+    def testV1_1Fp32CompareTorchEagerAgainstHuggingFace(self, huggingface_repo_id: str):
+        get_dataset(
+            huggingface_repo_id,
+        ).download()
```
Comment (IanNod) [inline sharktank/tests/models/t5/t5_test.py:56]: should we just cache the model for the CI instead of downloading each time? May end up with failures due to corrupted downloads on occasion.

File: sharktank/tests/ops/ops_test.py [hunk 10]
```diff
@@ -35,26 +34,6 @@ def testBroadcastDims(self):
         assert res[1] == 2
 
 
-class ElementwiseTest(unittest.TestCase):
```
Comment (IanNod) [inline sharktank/tests/ops/ops_test.py:38]: Why are we no longer testing?

File: .github/workflows/ci_eval.yaml (hunk 5 above)
Comment (sogartar) [reply .github/workflows/ci_eval.yaml:184]: It takes 30 seconds. These are 2 models small and xxl. Maybe I can leave only small on presubmit.

Comment (sogartar) [reply .github/workflows/ci_eval.yaml:184]: These presubmit tests will need to run on `llama-mi300x-3` or other machines that have the files available.

File: sharktank/sharktank/ops/default_impls.py (hunk 7 above)
Comment (KyleHerndon) [reply sharktank/sharktank/ops/default_impls.py:150]: I also expect this should be InferenceTensor

File: .github/workflows/ci_eval.yaml (hunk 5 above)
Comment (sogartar) [reply .github/workflows/ci_eval.yaml]: I refactored a bit. Added option `--with-t5-data` that enables tests that depend on T5 data. Made only the XXL eager mode test longrun. Added a presubmit job that runs the `small` model test.

File: sharktank/sharktank/ops/default_impls.py (hunk 7 above)
Comment (sogartar) [reply sharktank/sharktank/ops/default_impls.py]: Fixed the typo.
Originally I added this for convenience. Here is removed as it conflicts with dispatching to ternary ops. If the user wants binary op + folding they should do that themselves.

File: sharktank/tests/ops/ops_test.py (hunk 10 above)
Comment (sogartar) [reply sharktank/tests/ops/ops_test.py:38]: This tests the binary op + folding, which is no longer present.

File: sharktank/sharktank/ops/signatures.py (hunk 8 above)
Comment (sogartar) [reply sharktank/sharktank/ops/signatures.py:379]: Yes, this approximation is proposed in the paper.

File: .github/workflows/ci_eval.yaml [hunk 11]
```diff
@@ -20,10 +20,12 @@ concurrency:
   group: ${{ github.workflow }}-${{ github.event.number || github.sha }}
   cancel-in-progress: true
 
+# TODO: refactor out common steps into actions/scripts.
+
 jobs:
   test_perplexity_vmfb:
     timeout-minutes: 1000
-    name: "IREE/vmfb"
+    name: "IREE/vmfb perplexity"
```
Comment (archana-ramalingam) [inline .github/workflows/ci_eval.yaml:28]: Nit: I removed this earlier because it's `CI-Perplexity`.

File: sharktank/conftest.py [hunk 12]
```diff
@@ -133,6 +143,20 @@ def pytest_addoption(parser):
         help="Llama3.1 405b fp8 model path",
     )
 
+    parser.addoption(
+        "--google-t5-v1-1-small-fp32-model-path",
+        type=Path,
+        default="/data/t5/small/google__t5-v1_1-small_fp32.gguf",
```
Comment (archana-ramalingam) [inline sharktank/conftest.py:149]: We don't hardcode any llama model/tokenizer paths anymore here. You can pass it as an arg directly to pytest.

File: sharktank/sharktank/models/t5/t5.py (hunk 4 above)
Comment (archana-ramalingam) [reply sharktank/sharktank/models/t5/t5.py:574]: Yes, this was it. Are you able to test if removing this block made a diff in the results?

File: .github/workflows/ci_eval.yaml [hunk 13]
```diff
@@ -123,3 +125,63 @@ jobs:
...
+          # Try with the latest IREE nightly releases, not what iree-turbine pins.
+          # We could also pin to a known working or stable version.
+          # This should eventually stabilize. Do the best we can for now.
+          pip install -f https://iree.dev/pip-release-links.html --upgrade --pre \
+            iree-base-compiler \
+            iree-base-runtime
+
+      - name: Run long running tests
+        run: |
+          pytest
```
Comment (archana-ramalingam) [inline .github/workflows/ci_eval.yaml:184]: I need to check if the same perplexity dataset fits for both llama and T5.

File: sharktank/conftest.py (hunk 12 above)
Comment (sogartar) [reply sharktank/conftest.py:149]: Why not have a default that allows you to `pytest sharktank/tests` if you have data at default paths?

File: sharktank/sharktank/layers/ffn_block.py [hunk 14]
```diff
@@ -22,18 +23,29 @@ class FFN(ThetaLayer):
...
         self,
         theta: Theta,
+        is_gated: bool = True,
+        activation_fn: Optional[Callable[[AnyTensor], AnyTensor]] = None,
     ):
         super().__init__(theta)
 
-        self.add_module("ffn_gate", LinearLayer(theta("ffn_gate")))
+        self.is_gated = is_gated
+        self.activation_fn = activation_fn or F.silu
```
[the remark this answers is not in the document]
Comment (sogartar) [reply sharktank/sharktank/layers/ffn_block.py]: Done.

File: .github/workflows/ci_eval.yaml (hunk 5 above)
Comment (sogartar) [reply .github/workflows/ci_eval.yaml]: Done.

File: sharktank/sharktank/models/t5/t5.py (hunk 4 above)
Comment (sogartar) [reply sharktank/sharktank/models/t5/t5.py:574]: I have not done any float16 tests yet. Only float32 as this is the quantization of the original model from Google. I think this is an addition others made when doing fine tunning. I want to leave it in as this is also how inference works on the HF model.
This infinity probably comes from some previous calculation and if it appears it will blow up downstream results. It probably happens rarely.

File: .github/workflows/ci_eval.yaml (hunk 13 above)
Comment (sogartar) [reply .github/workflows/ci_eval.yaml]: I fixed the pytest command to be
```
pytest \
...
```

File: sharktank/tests/evaluate/perplexity_torch_test.py [hunk 15]
```diff
@@ -8,10 +8,9 @@
 import pytest
 import json
 
+from sharktank.utils.testing import longrun
```
Comment (archana-ramalingam) [inline sharktank/tests/evaluate/perplexity_torch_test.py:11]: I will add this change in my other PR, we can remove this here, if not using perplexity.

File: .github/workflows/ci_eval.yaml [hunk 16]
```diff
@@ -123,3 +125,63 @@ jobs:
 
       - name: Run perplexity test in eager mode
         run:  pytest -n 8 -v -s sharktank/tests/evaluate/perplexity_torch_test.py --longrun --llama3-8b-f16-model-path=/data/llama3.1/8b/llama8b_f16.irpa --llama3-8b-tokenizer-path=/data/llama3.1/8b/tokenizer_config.json
+
```
Comment (archana-ramalingam) [inline .github/workflows/ci_eval.yaml:128]: We can revert all changes to `ci_eval.yaml`, if not using perplexity.

File: sharktank/sharktank/layers/configs/llm_configs.py [hunk 17]
```diff
@@ -179,3 +179,72 @@ class LlamaModelConfig:
     # be the difference of many gigabytes of static data being embedded in
     # the program and not.
     static_tables: bool = True
+
+
+@dataclass
+class T5Config:
+    return_dict: bool = True
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/configs/llm_configs.py:186]: This is not being used elsewhere. Also, possible to stay consistent with the format `LlamaHParams` returns?

File: sharktank/sharktank/layers/ffn_block.py [hunk 18]
```diff
@@ -22,18 +23,29 @@ class FFN(ThetaLayer):
     def __init__(
         self,
         theta: Theta,
+        is_gated: bool = True,
+        activation_fn: Callable[[AnyTensor], AnyTensor] = F.silu,
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/ffn_block.py:27]: Since we have default values for both `is_gated` and `activation_fn`. Can we make them optional? Llama models use this default values, so shouldn't break anything.

File: sharktank/sharktank/layers/configs/llm_configs.py (hunk 17 above)
Comment (sogartar) [reply sharktank/sharktank/layers/configs/llm_configs.py:186]: It is possible to refactor out something the is common to Llama and T5. T5 does not have rotary embedding layer and and has `relative_buckets_count`. In general we could make dataclass hierarchy that reflects the GGUF properties. On the other hand there is already the properties dictionary if you want general. These config classes are tailored towards the model. There is some potential for code reuse. I would rather do that later.

File: sharktank/sharktank/layers/ffn_block.py (hunk 18 above)
Comment (sogartar) [reply sharktank/sharktank/layers/ffn_block.py:27]: Do you mean `Optional[SomeType]`?
`Optional[SomeType] == Union[SomeType, type(None)]`. They are not really optional in that sense as they can't be `None`. I had them optional in the first iteration, but you requested to not use `None` values for defaults.

File: .github/workflows/ci_eval.yaml (hunk 16 above)
Comment (sogartar) [reply .github/workflows/ci_eval.yaml]: Done.

File: sharktank/tests/evaluate/perplexity_torch_test.py (hunk 15 above)
Comment (sogartar) [reply sharktank/tests/evaluate/perplexity_torch_test.py]: I revered this change.

File: sharktank/sharktank/models/t5/t5.py [hunk 19]
```diff
@@ -0,0 +1,1067 @@
...
+        relative_attention_num_buckets: int,
+        relative_attention_max_distance: int,
+        d_model: int,
+        d_kv: int,
+        num_heads: int,
+        activation_dtype: torch.dtype,
+        has_relative_attention_bias: bool = False,
+    ):
+        super().__init__()
+        self.is_decoder = is_decoder
```
[the remark this answers is not in the document]
Comment (sogartar) [reply sharktank/sharktank/models/t5/t5.py:80]: Yes, I decided to leave this part in if/when we add the decoder.

File: .github/workflows/ci-sharktank.yml [hunk 20]
```diff
@@ -70,3 +70,63 @@ jobs:
         if: ${{ !cancelled() }}
         run: |
           pytest -n 4 sharktank/
+
+
+  test_with_data:
+    timeout-minutes: 1000
```
[the remark this answers is not in the document]
Comment (sogartar) [reply .github/workflows/ci-sharktank.yml]: I reduced it to 30 minutes.

File: sharktank/sharktank/layers/ffn_block.py (hunk 18 above)
Comment (archana-ramalingam) [reply sharktank/sharktank/layers/ffn_block.py:27]: Oh yeah, my bad, this shouldn't break llama. This looks good.

File: .github/workflows/ci-sharktank.yml (hunk 20 above)
[the remark this answers is not in the document]
Comment (IanNod) [reply .github/workflows/ci-sharktank.yml]: Does this take long enough that we need to specify a timeout here at all?

[the remark this answers is not in the document]
Comment (sogartar) [reply .github/workflows/ci-sharktank.yml]: You are right. I removed it. The default is 6 hours. Should be plenty. We should expect this to not stall.

**Review by IanNod — APPROVED** [review]

File: sharktank/sharktank/ops/default_impls.py (hunk 7 above)
Comment (sogartar) [reply sharktank/sharktank/ops/default_impls.py]: @KyleHerndon, I think this should not accept any inference tensor as we do unboxing, which would be undesirable for some tensor types. The unboxing here is fine as it is a no-op.
If someone wants an implementation for example for quantized tensors they should write one.
I guess it is about whether we want to perform potentially a very inefficient computation or to fail.

File: sharktank/conftest.py (hunk 12 above)
Comment (sogartar) [reply sharktank/conftest.py:158]: I added a comment.
